### ISSUE TITLE
Icecast support?

### ISSUE THREAD

**Issue by mattrei** [sebpiq/rhizome#113]
Hi,

I want to use rhizome together with SonicPi.
From SonicPi I send simply osc messages. 

However I also want to stream the Audio that is outputed from SonicPi to rhizome. I need the Icecast stack for that to work. 
Do you plan in adding it? If not maybe I can try to integrate it, but I am a newbie on that.
Are there other ways to do it?

best

**Comment by sebpiq** [comment]
Hi @mattrei 

I am not sure I understand your complete setup. The sonicPi would send OSC messages to what?
If you can tell more details, I can tell you how you could do it :)
However, rhizome doesn't deal with audio streaming, it just deals with message sending OSC, WebSockets, etc ... so a first answer would be that you should run your icecast server on the Pi, and add the stream address to your web page. But I don't understand (yet) where rhizome fits in this picture.

**Comment by mattrei** [comment]
Hi, maybe I do not understand the UseCase for Rhizome...

I am coding music with SonicPi. For my visuals I use WebGL in the browser. To drive the visuals I send OSC messages from sonicPi to the browser: however I can not send it directly but I need and OSC server between. Rhizome provides such an Interface (besides other cools stuff), so I have use for Rhizome ;)

However just sending OSC messages is a bit boring because I cannot send the audio data via OSC (at least it was not desigend for that). So my idea is to stream the audio via Icecast (where Rhizome provides the Icecast server besides OSC, Websockets, etc.) and stream it to my WebGL browser client. Then I can use the Icecast stream as my audio input and do all that fancy audio analysis which is possible with WebAudio.

Wouldn't that be a nice feature?

Best

**Comment by sebpiq** [comment]
Alright I see :)
Well then you can use rhizome as you said to send messages 
SonicPi [OSC] -> Web browser [WebSocket]
And also to serve the web page.
Then you can run the icecast server on the Pi, and stream directly audio that you generate there (although you should do some testing cause the Pi might not be powerful enough).
If your Pi has an assigned IP address (for example 192.168.2.45), then you can just add an audio tag (I don't remember the exact syntax) to your web page 

`<audio><source href="192.168.2.45:8000/myStream.ogg"> ... etc ...</audio>`

So basically, what I am trying to say, is that although you might use rhizome to send the messages and serve the web page, you don't need to integrate icecast with it. Is it clear?

This said, I hope you are conscious that when you stream audio you will have quite a lot of latency ... probably something between 10 to 30 seconds!!!

**Comment by mattrei** [comment]
You are the pro :+1: !! Would not have thought, that browser just support this out of the box.

Anyway you are right as well with the usability, its not built for low latency use cases. I get around 5 secs latency and therefore I have to throw this idea away (I am running all on localhost not on a RaspberryPI [SonicPI runs perfectly fine without it]).

Do you know of any (open-source) technology that can be used for my requirements, that means streaming audio from Supercollider to the browser?

Thanks

**Comment by sebpiq** [comment]
Hmm ... I received this question quite a few times, and I don't think it is possible. These technologies are not meant for low-latency streaming.

We have done this : https://vimeo.com/144312243
The source code is here https://github.com/sebpiq/fields 

The approach is different than streaming as we load WebPd patches in the browser of each phone. Maybe that could help!?
